File: Json/AT4_PythonGranularidade/crud.py
import json

# Menu Principal
def menu():

    print("\n=== MENU DE CONTATOS ===")
    print("1- Adicionar Contato")
    print("2- Listar Contatos")
    print("3- Atualizar Contato")
    print("4- Excluir")
    print("5- Sair")
    print("=========================\n")

# Função para escolher entre professores e alunos
def escolher_grupo():
    # Pergunta ao usuário o grupo que deseja retornar
    print("\nTipo de Contato: ")
    print("1 - Aluno")
    print("2 - Professor")

    opcao = input("Escolha: ")

    if opcao == "1":
        return "alunos"
    
    elif opcao == "2":
        return "professores"
    
    else:
        print("Opção Inválida!")

# Função para ler os dados do JSON
def ler_dados():
    with open("contatos.json", "r", encoding="utf-8") as arquivo:
        return json.load(arquivo)

# Função salvar dados
def salvar_dados(dados):
    with open("contatos.json", "w", encoding="utf-8") as arquivo:
        json.dump(dados, arquivo, indent=2, ensure_ascii=False)

# Adicionar dados
def adicionar():
    grupo = escolher_grupo()

    if not grupo:
        return
    
    nome = input("Nome: ")
    telefone = input("Telefone: ")

    dados = ler_dados()

    dados[grupo].append({
        "nome": nome,
        "telefone": telefone
    })
    
    salvar_dados(dados)
    print("Contato Adicionado!")

# Função listar contatos
def listar():
    grupo = escolher_grupo()

    if not grupo:
        return
    
    dados = ler_dados()
    print(f"\nLista de {grupo.upper()}: ")

    for index, contato in enumerate(dados[grupo], start=1): # Vai fazer com que o indice seja enumerado em cada grupo
        print(f"{index}. {contato['nome']} - {contato['telefone']}")

# Atualizar
def atualizar():
    grupo = escolher_grupo()

    if not grupo:
        return
    
    dados = ler_dados()

    index = int(input("Indice do contato: ")) -1

    if 0 <= index < len(dados[grupo]):
        nome = input("Novo Nome: ")
        telefone = input("Novo Telefone: ")

        dados[grupo][index] = {
            "nome": nome,
            "telefone": telefone
        }
        salvar_dados(dados)
        print("Contato Atualizado!")

    else: 
        print("Índice Inválido...")

# Função para excluir
def excluir():
    grupo = escolher_grupo()
    if not grupo:
        return
    
    dados = ler_dados()

    index = int(input("Indice do contato: ")) -1

    # Verificar se o index é válido
    if 0 <= index < len(dados[grupo]):
        # Remove o elemento da matriz
        dados[grupo].pop(index)

        salvar_dados(dados)
        print("Contato Excluído!")

    else:
        print("Indice Inválido!")

# Menu Principal
def main():
    while True:
        menu()
        opcao = input("Escolha: ")
        if opcao == "1":
            adicionar()

        elif opcao == "2":
            listar()

        elif  opcao == "3":
            atualizar()

        elif opcao == "4":
            excluir()

        elif opcao == "5":
            print("Saindo...")
            break
        else: 
            print("Opção Inválido!")

# Executa o programa
main()

File: Json/AT2_Granularidade/crud.js
const fs = require("fs");
const prompt = require("prompt-sync");

// Menu Principal
function menu() {
    console.log("\nMenu de Contato");
    console.log("1. Adicionar Contato");
    console.log("2. Listar Contatos");
    console.log("3. Atualizar Contato");
    console.log("4. Excluir Contato");
    console.log("5. Sair");
}

function escolherGrupo() {
    console.log("\nTipo de Contato");
    console.log("1. Aluno");
    console.log("2. Professor");

    const opcao = prompt("Escolha: ");

    // Ternária --> Simplificação de uma operação 
    if (opcao === "1") return "alunos";
    if (opcao === "2") return "professores";

    console.log("Opção Inválida!");
    return null;
}

// Leitura do JSON
function lerDados() {
    const dados = fs.readFileSync("contatos.json", "utf-8");
    return JSON.parse(dados);
}

function salvarDados(dados) {
    fs.writeFileSync("contatos.json", JSON.stringify(dados, null, 2));
}

// Função para adicionar
function adicionar() {
    const grupo = escolherGrupo();
    if (!grupo) return;

    const nome = prompt("Nome: ");
    const telefone = prompt("Telefone: ");

    const dados = lerDados();
    dados[grupo].push({nome, telefone});

    salvarDados(dados);
    console.log("Contato Adicionado!");
}

// Função para listar
function listar() {
    const grupo = escolherGrupo();
    if(!grupo) return;

    const dados = lerDados();
    console.log(`\nLista de ${grupo.toUpperCase()}: `);
    dados[grupo].forEach((contato, index) => {
        console.log(`${index + 1}. ${contato.nome} - ${contato.telefone}`);
    });
}

// Função atualizar 
function atualizar() {
    const grupo = escolherGrupo();
    if (!grupo) return;

    const dados = lerDados();
    const index = parseInt(prompt("Index do Contato: ")) -1;
    if (index >= 0 && index < dados[grupo].length){
        const nome = prompt("Novo nome: ");
        const telefone = prompt("Novo telefone: ");

        dados[grupo][index] = { nome, telefone };
        salvarDados(dados);

        console.log("Contato Atualizado!");
    }

    else {
        console.log("Índice Inválido!");
    }
}

// Função deletar
function excluir() {
    const grupo = escolherGrupo();
    if (!grupo) return;

    const dados = lerDados();
    const index = prseInt(prompt("Número do contato; ")) -1;

    if (index >= 0 && index < dados[grupo].length){
        dados[grupo].splice(index, 1);
        salvarDados(dados);

        console.log("Contato Excluído!");
    }

    else {
        console.log("Índice Inválido");
    }
}

// Programa principal
function main() {
    let opcao;

    do {
        menu();
        opcao = prompt("Escolha uma opção: ");
        switch (opcao){
            case "1":
                adicionar(); // Chama a função para adicionar
                break;
            case "2":
                listar(); // Chama a função listar
                break;
            case "3":
                atualizar(); // Chama a função atualizar
                break;
            case "4":
                excluir(); // Chama a função para excluir
                break;
            case "5":
                console.log("Saindo do programa..."); // Sai do programa
                break;
            default: // Caso o valor digitado não tiver opção
                console.log("Opção inválida. Tente novamente!");
        }
    } while (opcao !== "5");
}

main();